File: patch_setup.py
import re

with open('package/luci-app-setup/files/www/luci-static/resources/view/setup/setup.js', 'r') as f:
    content = f.read()

# I will add `var managedSids = {};` right before `applyRadioHtmode`
if "var managedSids = {" not in content:
    content = content.replace(
        "                applyRadioHtmode(radio2g, '2g', state);",
        "                var managedSids = { 'wizard_hotspot': true };\n\n                applyRadioHtmode(radio2g, '2g', state);"
    )

    # Record Hotspot Quick
    content = content.replace(
        "                                ensureNamedWifiIface(HOTSPOT_QUICK_IFACE_PRIMARY);",
        "                                ensureNamedWifiIface(HOTSPOT_QUICK_IFACE_PRIMARY);\n                                managedSids[HOTSPOT_QUICK_IFACE_PRIMARY] = true;"
    )
    content = content.replace(
        "                                ensureNamedWifiIface(HOTSPOT_QUICK_IFACE_SECONDARY);",
        "                                ensureNamedWifiIface(HOTSPOT_QUICK_IFACE_SECONDARY);\n                                managedSids[HOTSPOT_QUICK_IFACE_SECONDARY] = true;"
    )

    # Record Uplink AP
    content = content.replace(
        "                                uplinkLanApIface = ensureNamedWifiIface('wizard_uplink_ap');",
        "                                uplinkLanApIface = ensureNamedWifiIface('wizard_uplink_ap');\n                                managedSids[uplinkLanApIface] = true;"
    )

    # Record Mesh
    content = content.replace(
        "                        meshIface = ensureNamedWifiIface('wizard_mesh');",
        "                        meshIface = ensureNamedWifiIface('wizard_mesh');\n                        managedSids[meshIface] = true;"
    )
    content = content.replace(
        "                        meshApIface = meshRadio ? findWifiIface(meshRadio['.name']) : null;",
        "                        meshApIface = meshRadio ? findWifiIface(meshRadio['.name']) : null;\n                        if (meshApIface) managedSids[meshApIface] = true;"
    )

    # Record 2g LAN
    content = content.replace(
        "                        iface = ensureWifiIface(radio2g['.name']);",
        "                        iface = ensureWifiIface(radio2g['.name']);\n                        managedSids[iface] = true;"
    )
    # Record 2g VLAN
    content = content.replace(
        "                                ensureNamedWifiIface(secondaryIface2g);",
        "                                ensureNamedWifiIface(secondaryIface2g);\n                                managedSids[secondaryIface2g] = true;"
    )

    # Record 5g LAN
    content = content.replace(
        "                        iface = ensureWifiIface(radio5g['.name']);",
        "                        iface = ensureWifiIface(radio5g['.name']);\n                        managedSids[iface] = true;"
    )
    # Record 5g VLAN
    content = content.replace(
        "                                ensureNamedWifiIface(secondaryIface5g);",
        "                                ensureNamedWifiIface(secondaryIface5g);\n                                managedSids[secondaryIface5g] = true;"
    )

with open('/tmp/loop_to_replace.txt', 'r') as f:
    search_loop = f.read()

replace_loop = """                uci.sections('wireless', 'wifi-iface').forEach(function(section) {
                        var sid = section['.name'];
                        var sectionNetworks = normalizeList(section.network);
                        var isLocalRadio = localRadios.some(function(radio) {
                                return section.device == radio['.name'];
                        });

                        if (section.mode == null || section.mode == 'ap')
                                uci.set('wireless', sid, 'disassoc_low_ack', '0');

                        if (!isLocalRadio)
                                return;

                        if (section.mode != null && section.mode != 'ap')
                                return;

                        if (!managedSids[sid]) {
                                uci.remove('wireless', sid);
                                return;
                        }

                        if (sectionNetworks.indexOf('lan') > -1) {
                                if (vlanOnlyAp) {
                                        uci.remove('wireless', sid);
                                        return;
                                }

                                if (lanPolicy.enableWds)
                                        uci.set('wireless', sid, 'wds', '1');
                                else
                                        uci.unset('wireless', sid, 'wds');

                                applyWifiIfaceFlag('wireless', sid, 'hidden', lanPolicy.hidden);
                                applyWifiIfaceFlag('wireless', sid, 'isolate', lanPolicy.isolate);
                        }
                        else if (sectionNetworks.indexOf('wizardvlan') > -1) {
                                if (vlanPolicy.enableWds)
                                        uci.set('wireless', sid, 'wds', '1');
                                else
                                        uci.unset('wireless', sid, 'wds');

                                applyWifiIfaceFlag('wireless', sid, 'hidden', vlanPolicy.hidden);
                                applyWifiIfaceFlag('wireless', sid, 'isolate', vlanPolicy.isolate);
                        }
                });"""

content = content.replace(search_loop, replace_loop)

print("Replaced:", search_loop in content, replace_loop in content)
with open('package/luci-app-setup/files/www/luci-static/resources/view/setup/setup.js', 'w') as f:
    f.write(content)

File: patch_setup.js
const fs = require('fs');
let code = fs.readFileSync('package/luci-app-setup/files/www/luci-static/resources/view/setup/setup.js', 'utf8');

// Ensure we don't apply twice
if (code.includes('var managedSids = {}')) {
    console.log('Already applied');
    process.exit(0);
}

// 1. Add variable definition
code = code.replace(
    /(applyRadioHtmode\(radio2g[^;]+;\s*)/,
    "var managedSids = {};\n\t\t$1"
);

// 2. Track Hotspot Quick interfaces
// ensureNamedWifiIface(HOTSPOT_QUICK_IFACE_PRIMARY);
code = code.replace(
    /(ensureNamedWifiIface\(HOTSPOT_QUICK_IFACE_PRIMARY\);)/g,
    "$1\n\t\t\t\tmanagedSids[HOTSPOT_QUICK_IFACE_PRIMARY] = true;"
);
code = code.replace(
    /(ensureNamedWifiIface\(HOTSPOT_QUICK_IFACE_SECONDARY\);)/g,
    "$1\n\t\t\t\tmanagedSids[HOTSPOT_QUICK_IFACE_SECONDARY] = true;"
);

// 3. Track Uplink AP
// uplinkLanApIface = ensureNamedWifiIface('wizard_uplink_ap');
code = code.replace(
    /(uplinkLanApIface\s*=\s*ensureNamedWifiIface\('wizard_uplink_ap'\);)/g,
    "$1\n\t\t\t\t\tmanagedSids[uplinkLanApIface] = true;"
);

// 4. Track Mesh
// meshIface = ensureNamedWifiIface('wizard_mesh');
code = code.replace(
    /(meshIface\s*=\s*ensureNamedWifiIface\('wizard_mesh'\);)/g,
    "$1\n\t\t\t\t\tmanagedSids[meshIface] = true;"
);
// meshApIface = meshRadio ? findWifiIface(meshRadio['.name']) : null;
code = code.replace(
    /(meshApIface\s*=\s*meshRadio\s*\?\s*findWifiIface[^;]+;)/g,
    "$1\n\t\t\t\t\t\tif (meshApIface) managedSids[meshApIface] = true;"
);

// 5. Track 2g LAN
// iface = ensureWifiIface(radio2g['.name']);
code = code.replace(
    /(iface\s*=\s*ensureWifiIface\(radio2g\['\.name'\]\);)/g,
    "$1\n\t\t\t\t\t\tmanagedSids[iface] = true;"
);
// Track 2g VLAN
// ensureNamedWifiIface(secondaryIface2g);
code = code.replace(
    /(ensureNamedWifiIface\(secondaryIface2g\);)/g,
    "$1\n\t\t\t\t\t\t\tmanagedSids[secondaryIface2g] = true;"
);

// 6. Track 5g LAN
// iface = ensureWifiIface(radio5g['.name']);
code = code.replace(
    /(iface\s*=\s*ensureWifiIface\(radio5g\['\.name'\]\);)/g,
    "$1\n\t\t\t\t\t\tmanagedSids[iface] = true;"
);
// Track 5g VLAN
// ensureNamedWifiIface(secondaryIface5g);
code = code.replace(
    /(ensureNamedWifiIface\(secondaryIface5g\);)/g,
    "$1\n\t\t\t\t\t\t\tmanagedSids[secondaryIface5g] = true;"
);

// 7. Modifying the loop
const loopRegex = /uci\.sections\('wireless',\s*'wifi-iface'\)\.forEach\([\s\S]*?\}\);/;
const replaceLoop = `uci.sections('wireless', 'wifi-iface').forEach(function(section) {
\t\t\tvar sid = section['.name'];
\t\t\tvar sectionNetworks = normalizeList(section.network);
\t\t\tvar isLocalRadio = localRadios.some(function(radio) {
\t\t\t\treturn section.device == radio['.name'];
\t\t\t});

\t\t\tif (section.mode == null || section.mode == 'ap')
\t\t\t\tuci.set('wireless', sid, 'disassoc_low_ack', '0');

\t\t\tif (!isLocalRadio)
\t\t\t\treturn;

\t\t\tif (section.mode != null && section.mode != 'ap')
\t\t\t\treturn;

\t\t\tif (!managedSids[sid]) {
\t\t\t\tuci.remove('wireless', sid);
\t\t\t\treturn;
\t\t\t}

\t\t\tif (sectionNetworks.indexOf('lan') > -1) {
\t\t\t\tif (vlanOnlyAp) {
\t\t\t\t\tuci.remove('wireless', sid);
\t\t\t\t\treturn;
\t\t\t\t}

\t\t\t\tif (lanPolicy.enableWds)
\t\t\t\t\tuci.set('wireless', sid, 'wds', '1');
\t\t\t\telse
\t\t\t\t\tuci.unset('wireless', sid, 'wds');

\t\t\t\tapplyWifiIfaceFlag('wireless', sid, 'hidden', lanPolicy.hidden);
\t\t\t\tapplyWifiIfaceFlag('wireless', sid, 'isolate', lanPolicy.isolate);
\t\t\t}
\t\t\telse if (sectionNetworks.indexOf('wizardvlan') > -1) {
\t\t\t\tif (vlanPolicy.enableWds)
\t\t\t\t\tuci.set('wireless', sid, 'wds', '1');
\t\t\t\telse
\t\t\t\t\tuci.unset('wireless', sid, 'wds');

\t\t\t\tapplyWifiIfaceFlag('wireless', sid, 'hidden', vlanPolicy.hidden);
\t\t\t\tapplyWifiIfaceFlag('wireless', sid, 'isolate', vlanPolicy.isolate);
\t\t\t}
\t\t});`;

code = code.replace(loopRegex, replaceLoop);

fs.writeFileSync('package/luci-app-setup/files/www/luci-static/resources/view/setup/setup.js', code);
console.log('Update successful');
